marimbabot_audio/scripts/onset_detection.py

```diff
@@ -0,0 +1,547 @@
+#!/usr/bin/env python
+import cv_bridge
+import rospy
+from audio_common_msgs.msg import AudioDataStamped, AudioInfo
+from sensor_msgs.msg import Image
+from std_msgs.msg import Float32, ColorRGBA
+from marimbabot_audio.msg import NoteOnset, CQTStamped
+
+
+import pretty_midi
+import librosa
+import crepe
+import crepe.core
+
+import cv2
+import numpy as np
+import matplotlib.pyplot as plt
+from matplotlib.colors import ListedColormap
+from functools import reduce
+import struct
+
+
+# import time
+'''
+    Some default parameters:
+        -   Sampling rate = 44100
+        -   Sampleing format = S16LE
+
+'''
+def hz_to_note(hz):
+    return pretty_midi.note_number_to_name(pretty_midi.hz_to_note_number(hz))
+def note_to_pm_id(note_name):
+    return pretty_midi.note_name_to_number(note_name)
+def pm_id_to_note(pm_id):
+    return pretty_midi.note_number_to_name(pm_id)
+def note_to_hz(note):
+    return pretty_midi.note_number_to_hz(pretty_midi.note_name_to_number(note))
+
+def check_audio_format():
+    """
+    check available audio input format, if it satisfy the requirements.
+    :return:
+    """
+    rospy.loginfo("Waiting for Audio Info")
+    info = rospy.wait_for_message("audio_info", AudioInfo)
+    if info.channels != 1:
+        rospy.logfatal(
+            "audio data has more than one channel,"
+            "expecting single-channel recording"
+        )
+    elif info.sample_rate != 44100:
+        rospy.logfatal(
+            f"sample rate {info.sample_rate} is not 44100"
+        )
+    elif info.sample_format != "S16LE":
+        rospy.logfatal(
+            f"sample format '{info.sample_format}' is not S16LE"
+        )
+    elif info.coding_format != "wave":
+        rospy.logfatal(
+            f"coding '{info.coding_format}' is not raw"
+        )
+    else:
+        rospy.loginfo("Audio compatible")
+        return True
+    return False
+
+def unpack_data(data):
+    return np.array(struct.unpack(
+        f"{int(len(data) / 2)}h",
+        bytes(data)
+        ),
+        dtype=float)
+
+
+class OnsetDetection:
+    def _init_instrument_config(self):
+        # # harp
+        # self.fmin_note = "C4"
+        # self.fmax_note = "C6"
+        # self.semitones = 62
+
+        # # guzheng
+        # self.fmin_note = "C2"
+        # self.fmax_note = "C8"
+        # self.semitones = 84
+
+        # marimba
+        self.fmin_note = "C4"  # C4 pm_id=60
```

Comment:
You might want to use ros parameters for things like that.

Reply:
Have a look at the vision if you have further questions.

Reply:
those parameters are just for our marimba instrument.